File: src/main/java/nz/co/searchwellington/htmlparsing/CompositeLinkExtractor.java
package nz.co.searchwellington.htmlparsing;

import java.util.HashSet;
import java.util.Set;

import com.google.common.collect.Sets;
import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class CompositeLinkExtractor implements LinkExtractor {

    private LinkExtractor[] linkExtractors;
    
    @Autowired
    public CompositeLinkExtractor(LinkExtractor[] linkExtractors) {
		this.linkExtractors = linkExtractors.clone();
	}

	public Set<String> extractLinks(String inputHTML) {
        Set<String> links = Sets.newHashSet();
        if (inputHTML != null) {
        	for (LinkExtractor linkExtractor : linkExtractors) {
                links.addAll(linkExtractor.extractLinks(inputHTML));
			}
        }
        return links;
    }

}

File: src/main/scala/nz/co/searchwellington/htmlparsing/CompositeLinkExtractor.scala
package nz.co.searchwellington.htmlparsing

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component class CompositeLinkExtractor @Autowired()(linkExtractors: Array[LinkExtractor]) extends LinkExtractor {

  override def extractLinks (inputHTML: String): Seq[String] = {
    linkExtractors.toSeq.flatMap(_.extractLinks(inputHTML))
  }

}